http-webhook.md

```diff
@@ -0,0 +1,339 @@
+4. [Abuse Protection](#4-abuse-protection)
+5. [References](#5-references)
+
+## 1. Introduction
+
+"Web hooks" are a popular pattern to deliver notifications between applications
+and via HTTP endpoints. Applications that make notifications available, allow
+for other applications to register an HTTP endpoint to which notifications are
+delivered once available.
+
+This specification defines a HTTP method by how notifications are delivered by
+the sender, an authorization model for event delivery to protect the receiver,
+and a registration handshake that protects the sender from being abused for
+flooding arbitrary HTTP sites with requests.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][RFC2119].
+
+### 1.2. Relation to HTTP
+
+This specification prescribes rules constraining the use and handling of
+specific [HTTP methods][RFC7231-Section-4] and headers.
+
+This specification also applies equivalently to HTTP/2 ([RFC7540][RFC7540]),
+which is compatible with HTTP 1.1 semantics.
+
+## 2. Delivering notifications
+
+### 2.1. Delivery request
+
+Notifications are delivered using a HTTP request. The response indicates the
+resulting status of the delivery.
+
+HTTP-over-TLS (HTTPS) [RFC2818][RFC2818] MUST be used for the connection.
+
+The HTTP method for the delivery request MUST be [POST][POST].
+
+The [`Content-Type`][Content-Type] header MUST be carried and the request MUST
+carry a notification payload of the given content type. Requests without
+payloads, e.g. where the notification is entirely expressed in HTTP headers,
+are not permitted by this specification.
+
+This specification does not further constrain the content of the notification,
+and it also does not prescribe the [HTTP target resource][RFC7230-Section-5-1]
+that is used for delivery.
+
+If the delivery target supports and requires [Abuse Protection][#4-abuse-protection],
+the delivery request MUST include the `Origin` header. The `Origin` header value
+is a DNS name expression that identifies the sending system.
+
+### 2.2. Delivery response
+
+The delivery response MAY contain a payload providing detail status
+information in the case of handling errors. This specification does not attempt
+to define such a payload.
+
+The response MUST NOT use any of the [3xx HTTP Redirect status codes][3xx] and
+the client MUST NOT follow any such redirection.
+
+If a delivery target has been retired, but the HTTP site still exists, the
+site SHOULD return a [410 "Gone"][410] status code and the sender SHOULD refrain
+from sending any further notifications.
+
+If the delivery target is unable to process the request due to exceeding a
+request rate limit, it SHOULD return a [429 Too Many Requests][429] status code
+and MUST include the [`Retry-After`][Retry-After] header. The sender MUST
```

Comment:
This may be nitty, though I'm curious whether a distributed sender MUST coordinate/interlock so that whenever one shard gets a 429, all shards must respect that immediately.

Reply:
I don't think it's realistic or necessary to expect perfect consistency across multiple nodes that act as if they were one sender. If there are 4 concurrent requests in flight from 4 nodes, each of them is presumably going to get a 429 each and will subsequently back off. For the protocol, multiple senders acting as one doesn't really exist. You'll be smart to have an eventually consistent gossip (or other mechanism) to tell your push agents about a target currently throttling, so that you can move on to other work while the condition persists. That is ultimately what Retry-After helps with.